Add bulk publish documentation (#3005)
dapr/docs: daprdocs/content/en/reference/api/pubsub_api.md

> Additional metadata parameters are available based on each pubsub component.

## Publish multiple messages to a given topic

This endpoint lets you publish multiple messages to consumers who are listening on a `topic`.

### HTTP Request

```
POST http://localhost:<daprPort>/v1.0-alpha1/publish/bulk/<pubsubname>/<topic>[?<metadata>]
```

> The request body should contain a JSON array of entries with unique entry IDs, the event to publish, and the content type of the event. If the content type for an event is not `application/cloudevents+json`, it is auto-wrapped as a CloudEvent (unless `metadata.rawPayload` is set to `true`, see below).

Reviewer: ```suggestion
The request body should contain a JSON array of entries with:
- Unique entry IDs
- The event to publish
- The content type of the event

If the content type for an event is not `application/cloudevents+json`, it is auto-wrapped as a CloudEvent (unless `metadata.rawPayload` is set to `true`).
```

Example:

```bash
curl -X POST http://localhost:3500/v1.0-alpha1/publish/bulk/pubsubName/deathStarStatus \
  -H 'Content-Type: application/json' \
  -d '[
        {
            "entryId": "ae6bf7c6-4af2-11ed-b878-0242ac120002",
            "event":  "first",
            "contentType": "text/plain"
        },
        {
            "entryId": "b1f40bd6-4af2-11ed-b878-0242ac120002",
            "event":  {
                "message": "second"   
            },
            "contentType": "application/json"
        },
      ]'
```

### Headers

The `Content-Type` header should always be set to `application/json`.

### URL Parameters

|**Parameter**|**Description**|
|--|--|
|`daprPort`|The Dapr port|

> |`pubsubname`|The name of pubsub component|

Reviewer: ```suggestion
|`pubsubname`|The name of pub/sub component|
```

|`topic`|The name of the topic|

> |`metadata`|Query parameters for metadata as described below|

Reviewer: ```suggestion
|`metadata`|Query parameters for [metadata]({{< ref "pubsub_api.md#metadata" >}})|
```

### Metadata

> Metadata can be sent via query parameters in the request's URL. If must be prefixed with `metadata.` as shown below.

Reviewer: ```suggestion
Metadata can be sent via query parameters in the request's URL. It must be prefixed with `metadata.`, as shown in the table below.
```

|**Parameter**|**Description**|
|--|--|
|`metadata.rawPayload`|Boolean to determine if Dapr should publish the messages without wrapping them as CloudEvent.|
|`metadata.maxBulkPubBytes`|Maximum bytes to publish in a bulk publish request.|

#### HTTP Response

|**HTTP Status**|**Description**|
|--|--|
|204|All messages delivered|

> |400|Pubsub does not exist|

Reviewer: ```suggestion
|400|Pub/sub does not exist|
```

> Additional metadata parameters are available based on each pubsub component.

## Publish multiple messages to a given topic

This endpoint lets you publish multiple messages to consumers who are listening on a `topic`.

### HTTP Request

```
POST http://localhost:<daprPort>/v1.0-alpha1/publish/bulk/<pubsubname>/<topic>[?<metadata>]
```

The request body should contain a JSON array of entries with:
- Unique entry IDs
- The event to publish
- The content type of the event

If the content type for an event is not `application/cloudevents+json`, it is auto-wrapped as a CloudEvent (unless `metadata.rawPayload` is set to `true`).

Example:

```bash
curl -X POST http://localhost:3500/v1.0-alpha1/publish/bulk/pubsubName/deathStarStatus \
  -H 'Content-Type: application/json' \
  -d '[
        {
            "entryId": "ae6bf7c6-4af2-11ed-b878-0242ac120002",

>             "event":  "first",

Reviewer 2: ```suggestion
            "event":  "first text message",
```

            "contentType": "text/plain"
        },
        {
            "entryId": "b1f40bd6-4af2-11ed-b878-0242ac120002",
            "event":  {

>                 "message": "second"   

Reviewer 2: ```suggestion
                "message": "second JSON message"   
```

            },
            "contentType": "application/json"
        },
      ]'
```

### Headers

> The `Content-Type` header should always be set to `application/json`.

Reviewer 2: You have a "contentType": "text/plain" is the example for the first message. Is this correct?

Author: Yes, this is the request header, the request body itself has to be a valid JSON in the format specified. Data has a separate content-type, embedded in the request body.

Author: Added a few words to make this clearer.

### URL Parameters

|**Parameter**|**Description**|
|--|--|
|`daprPort`|The Dapr port|
|`pubsubname`|The name of pub/sub component|
|`topic`|The name of the topic|
|`metadata`|Query parameters for [metadata]({{< ref "pubsub_api.md#metadata" >}})|

### Metadata

Metadata can be sent via query parameters in the request's URL. It must be prefixed with `metadata.`, as shown in the table below.

|**Parameter**|**Description**|
|--|--|
|`metadata.rawPayload`|Boolean to determine if Dapr should publish the messages without wrapping them as CloudEvent.|
|`metadata.maxBulkPubBytes`|Maximum bytes to publish in a bulk publish request.|

#### HTTP Response

|**HTTP Status**|**Description**|
|--|--|
|204|All messages delivered|
|400|Pub/sub does not exist|
|403|Forbidden by access controls|
|500|At least one message failed to be delivered|

> The response body is a JSON containing a list of failed entries. Example:

Reviewer 2: Can you make is clear here. Do you get a list of the message that failed to deliver or just an error. If the former, I would use the example above and say, if the second JSON message failed to send (the one I have called "second JSON message") do you see the same entryId? I guess I am confused by the "error message" also, will this always be error message? Would help to explain this in some more details and using the same example as the successful send message above ^^

Author: > If the former, I would use the example above and say, if the second JSON message failed to send (the one I have called "second JSON message") do you see the same entryId?

Yes, added this explanation.

> will this always be error message?

WDYM? This should contain error messages from the underlying broker.
